Flaky DVC fetch `*.dvc: [WinError 5] Access is denied` errors running fetch_sources.py on Azure Windows build runners

Example failing workflow runs:
* https://github.com/ROCm/TheRock/actions/runs/27720997549/job/82009127320?pr=5933#step:14:140
* https://github.com/ROCm/TheRock/actions/runs/27704637735/job/81953351574#step:14:143
* https://github.com/ROCm/rockrel/actions/runs/27657263001/job/81811991074#step:14:138

```
Traceback (most recent call last):
dvc config detected in C:\home\runner\_work\TheRock\TheRock\rocm-libraries, fetching large files
pull: 20 pointer file(s) from s3://therock-dvc/rocm-libraries
  fetching  Gqa.tensor0.bin (512.0 KiB)
  fetching  bnorm_backward.tar.gz (1.3 KiB)
  fetching  MIOpen.tensor0.bin (96.0 B)
  fetching  Medium.tensor0.bin (1.0 MiB)
  fetching  Small.tensor0.bin (240.0 B)
  fetching  Small.tensor0.bin (480.0 B)
  fetching  Small.tensor0.bin (240.0 B)
  fetching  Small.tensor0.bin (512.0 KiB)
  fetching  Small.tensor0.bin (2.8 KiB)
  fetching  bnorm_fwd.tar.gz (1.3 KiB)
  fetching  conv_dgrad.tar.gz (9.3 KiB)
  fetching  Large.tensor0.bin (117.2 KiB)
  fetching  Small.tensor1.bin (12.0 B)
  fetching  conv_fwd.tar.gz (10.6 KiB)
  fetching  conv_wgrad.tar.gz (10.4 KiB)
  fetching  Small.tensor1.bin (12.0 B)
  fetching  Small.tensor1.bin (12.0 B)
  fetching  MIOpen.tensor1.bin (8.0 B)
  fetching  Small.tensor1.bin (12.0 B)
  fetching  gfx1030.kdb.bz2 (381.5 MiB)
  fetching  Small.tensor1.bin (512.0 KiB)
  fetching  Large.tensor1.bin (16.0 B)
  fetching  Small.tensor2.bin (12.0 B)
  fetching  Small.tensor2.bin (12.0 B)
  fetching  gfx900.kdb.bz2 (270.1 MiB)
  fetching  Small.tensor2.bin (12.0 B)
  fetching  gfx906.kdb.bz2 (318.4 MiB)
  fetching  Gqa.tensor1.bin (128.0 KiB)
  fetching  Small.tensor2.bin (12.0 B)
  fetching  MIOpen.tensor2.bin (8.0 B)
  File "C:\home\runner\_work\TheRock\TheRock\build_tools\fetch_sources.py", line 992, in <module>
    main(sys.argv[1:])
  File "C:\home\runner\_work\TheRock\TheRock\build_tools\fetch_sources.py", line 988, in main
    run(args)
  File "C:\home\runner\_work\TheRock\TheRock\build_tools\fetch_sources.py", line 536, in run
  fetching  Medium.tensor1.bin (1.0 MiB)
  fetching  Large.tensor2.bin (16.0 B)
  fetching  Small.tensor3.bin (12.0 B)
  fetching  Small.tensor3.bin (12.0 B)
  fetching  Small.tensor3.bin (12.0 B)
  fetching  Small.tensor2.bin (512.0 KiB)
  fetching  Small.tensor4.bin (12.0 B)
  fetching  gfx908.kdb.bz2 (104.7 MiB)
  fetching  Small.tensor5.bin (240.0 B)
  cached    Small.tensor3.bin (12.0 B)
  cached    Small.tensor4.bin (12.0 B)
  fetching  Small.tensor5.bin (480.0 B)
  fetching  MIOpen.tensor3.bin (8.0 B)
  fetching  gfx90a.kdb.bz2 (147.3 MiB)
  fetching  Large.tensor3.bin (16.0 B)
  fetching  Small.tensor4.bin (12.0 B)
  fetching  MIOpen.tensor4.bin (8.0 B)
  fetching  Medium.tensor2.bin (1.0 MiB)
  fetching  Gqa.tensor2.bin (128.0 KiB)
  fetching  gfx942.kdb.bz2 (3.2 MiB)
  fetching  Small.tensor3.bin (512.0 KiB)
  fetching  Large.tensor4.bin (16.0 B)
  fetching  Small.tensor5.bin (2.8 KiB)
  fetching  MIOpen.tensor5.bin (96.0 B)
  fetching  Gqa.tensor3.bin (512.0 KiB)
  fetching  Medium.tensor3.bin (1.0 MiB)
  fetching  Large.tensor5.bin (117.2 KiB)
  ERROR     C:\home\runner\_work\TheRock\TheRock\rocm-libraries\dnn-providers\integration-tests\golden_reference_data\quick\BatchnormFwdInference\nchw\fp16\Small\Small.tensors.dvc: [WinError 5] Access is denied: 'C:\\Users\\ContainerAdministrator\\.cache\\therock-dvc\\ce\\e73630e22266547f7e77a86560f21b.bd6b58c6fc044122a88a638b8bd368e2.tmp' -> 'C:\\Users\\ContainerAdministrator\\.cache\\therock-dvc\\ce\\e73630e22266547f7e77a86560f21b'
    pull_large_files(args.dvc_projects, projects, jobs=args.jobs)
  File "C:\home\runner\_work\TheRock\TheRock\build_tools\fetch_sources.py", line 571, in pull_large_files
    result = fetch_dvc_artifacts.pull(project_dir, jobs=pull_jobs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\home\runner\_work\TheRock\TheRock\build_tools\fetch_dvc_artifacts.py", line 527, in pull
    raise FetchError(f"{len(errors)} file(s) failed to fetch (see log above)")
fetch_dvc_artifacts.FetchError: 1 file(s) failed to fetch (see log above)
Error: Process completed with exit code 1.
```

Callstack points to somewhere in https://github.com/ROCm/TheRock/blob/main/build_tools/fetch_dvc_artifacts.py

Only seeing this on Windows, not Linux. Enough occurrences today to be noticeable, possibly increasing in frequency ... maybe more DVC files were added recently and pulled in via a submodule update?

--- a/build_tools/fetch_dvc_artifacts.py
+++ b/build_tools/fetch_dvc_artifacts.py
@@ -259,7 +259,16 @@ def _store_in_cache(src: Path, cache_file: Path) -> None:
             os.link(src, tmp)
         except OSError:
             shutil.copy2(src, tmp)
-        os.replace(tmp, cache_file)
+        try:
+            os.replace(tmp, cache_file)
+        except PermissionError:
+            # Windows raises PermissionError when two workers race to os.replace
+            # the same cache_file simultaneously (POSIX permits this silently).
+            # If the winner already wrote the correct content, treat as success.
+            if cache_file.exists() and src.stat().st_size == cache_file.stat().st_size:
+                if _md5_of(src) == _md5_of(cache_file):
+                    return
+            raise
     finally:
         if tmp.exists():
             try:

--- a/build_tools/tests/fetch_dvc_artifacts_test.py
+++ b/build_tools/tests/fetch_dvc_artifacts_test.py
@@ -15,6 +15,7 @@ import io
 import os
 import sys
 import tempfile
+import threading
 import unittest
 from pathlib import Path
 from unittest import mock
@@ -401,6 +402,81 @@ class TestDownloadBlob(unittest.TestCase):
             with self.assertRaisesRegex(fda.FetchError, "size mismatch"):
                 fda._download_blob(s3, remote, md5, dest, expected_size=len(data) + 100)
             self.assertFalse(dest.exists())
+
+
+class TestStoreInCacheConcurrency(unittest.TestCase):
+    """Regression test for #5943.
+
+    Multiple .dvc pointer files can reference files with the same content hash
+    (e.g. identical small tensors in different subdirectories). When fetched
+    concurrently, all workers call _store_in_cache with the same cache_file
+    path. Each uses a unique temp name, but the final os.replace races on
+    Windows and raises PermissionError. The fix catches PermissionError and
+    verifies the winner wrote the correct content before treating it as success.
+    """
+
+    def test_concurrent_calls_with_same_cache_file_do_not_raise(self):
+        with tempfile.TemporaryDirectory() as t:
+            tmp = Path(t)
+            data = b"shared content"
+            md5 = _md5_hex(data)
+            cache_file = fda._cache_path(tmp / "cache", md5)
+
+            # Simulate the Windows race: both workers arrive at os.replace
+            # concurrently. Caller 1 writes cache_file; caller 2 then gets
+            # PermissionError because Windows won't replace a file another
+            # handle still has open. We model this with a two-phase barrier:
+            # both workers reach the gate, then caller 1 does the real replace
+            # and only after that does caller 2 raise PermissionError, so
+            # cache_file already exists when the exception handler checks it.
+            original_replace = os.replace
+            call_count = [0]
+            lock = threading.Lock()
+            gate = threading.Barrier(2)
+            done = threading.Event()
+
+            def patched_replace(src, dst):
+                with lock:
+                    call_count[0] += 1
+                    n = call_count[0]
+                gate.wait(timeout=5)
+                if n == 1:
+                    original_replace(src, dst)
+                    done.set()
+                else:
+                    done.wait(timeout=5)
+                    raise PermissionError("[WinError 5] Access is denied")
+
+            src1 = tmp / "src1.bin"
+            src2 = tmp / "src2.bin"
+            src1.write_bytes(data)
+            src2.write_bytes(data)
+
+            errors = []
+
+            def worker(src):
+                try:
+                    fda._store_in_cache(src, cache_file)
+                except Exception as e:
+                    errors.append(e)
+
+            # Patch fda's reference to os.replace (not the os module itself)
+            # so both threads see the same controlled replacement.
+            with mock.patch.object(fda.os, "replace", side_effect=patched_replace):
+                threads = [
+                    threading.Thread(target=worker, args=(src1,)),
+                    threading.Thread(target=worker, args=(src2,)),
+                ]
+                for th in threads:
+                    th.start()
+                for th in threads:
+                    th.join()
+
+            self.assertEqual(errors, [], f"unexpected errors: {errors}")
+            self.assertTrue(cache_file.exists())
+            self.assertEqual(cache_file.read_bytes(), data)
+            leftover = list(cache_file.parent.glob("*.tmp"))
+            self.assertEqual(leftover, [], f"leftover temp files: {leftover}")
 
 
 class TestMaterializeFile(unittest.TestCase):

